pkg/kube/dialer.go

```diff
@@ -0,0 +1,336 @@
+	clusterDomain string
+	requestID     int
+
+	connections map[types.NamespacedName]httpstream.Connection
+	mutex       sync.RWMutex
+}
+
+// NewPodDialer create a PodDialer.
+func NewPodDialer(config *rest.Config) *PodDialer {
+	return &PodDialer{
+		config:        config,
+		clusterDomain: defaultClusterDomain,
+		connections:   make(map[types.NamespacedName]httpstream.Connection),
+	}
+}
+
+// WithClusterDomain overrides the domain used for FQDN parsing.
+func (p *PodDialer) WithClusterDomain(domain string) *PodDialer {
+	p.clusterDomain = domain
+	return p
+}
+
+// Dial dials the given pod's service-based DNS address and returns a
+// net.Conn that can be used to reach the pod directly.
+func (p *PodDialer) Dial(network string, address string) (net.Conn, error) {
+	switch network {
+	case "tcp", "tcp4", "tcp6":
+	default:
+		return nil, fmt.Errorf("dialer only supports TCP-based networks: %w", net.UnknownNetworkError(network))
+	}
+
+	pod, port, err := p.parseDNS(address)
+	if err != nil {
+		return nil, err
+	}
+
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	conn, err := p.connectionForPod(pod)
+	if err != nil {
+		return nil, err
+	}
+
+	p.requestID++
+
+	headers := http.Header{}
+	headers.Set(corev1.PortHeader, strconv.Itoa(port))
+	headers.Set(corev1.PortForwardRequestIDHeader, strconv.Itoa(p.requestID))
+
+	headers.Set(corev1.StreamType, corev1.StreamTypeError)
+	errorStream, err := conn.CreateStream(headers)
+	if err != nil {
+		return nil, err
+	}
+
+	headers.Set(corev1.StreamType, corev1.StreamTypeData)
+	dataStream, err := conn.CreateStream(headers)
+	if err != nil {
+		return nil, err
+	}
+
+	return wrapConn(network, address, dataStream, errorStream), nil
+}
+
+// DialContext is a simple wrapper around Dial.
+func (p *PodDialer) DialContext(_ context.Context, network string, address string) (net.Conn, error) {
+	return p.Dial(network, address)
+}
+
+// parseDNS attempts to determine the intended pod to target, currently the
+// following formats are supported:
+//  1. [pod-name].[service-name].[namespace-name].svc.[cluster-domain]
+//  2. [pod-name].[service-name].[namespace-name].svc
+//  3. [pod-name].[service-name].[namespace-name]
+//  4. [pod-name].[service-name]
+//
+// If no cluster-domain is supplied, the dialer's configured domain is assumed.
+// If no namespace-name is supplied, the default namespace is assumed.
+// No validation is done to ensure that the pod is actually referenced by the
+// given service or that the DNS record exists in Kubernetes, instead this assumes
+// that things are set up properly in Kubernetes such that the FQDN passed here
+// matches Kubernetes' own DNS records and a pod within the cluster would be able
+// to resolve the same FQDN to the pod that we do.
+//
+// These assumptions allow us to use this dialer to issues requests *as if* we are
+// in the Kubernetes network even from outside of it (i.e. in tests that attempt to
+// connect to a pod at a given DNS address).
```

Comment:
super nit: wouldn't the correct term be "given domain" or "given hostname"? I understand what you mean by DNS address but AFAIK it's not technically a correct term.